[RFC] Fix jemalloc assertion due to non-monotonic CLOCK_MONOTONIC_COARSE

azat opened the pull request:
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Fix jemalloc assertion due to non-monotonic CLOCK_MONOTONIC_COARSE

Recently one tricky assertion of jemalloc had been discovered [1]:

    Failed assertion: "nstime_compare(&decay->epoch, new_time) <= 0"

  [1]: https://github.com/ClickHouse/ClickHouse/issues/66193

And as it turns out it is really possible for CLOCK_MONOTONIC_COARSE to go backwards, in a nutshell it can be done with ADJ_FREQUENCY, you can find example here [2]. And I can't trigger this issue for non-coarse clocks.

  [2]: https://gist.github.com/azat/7ea7f50ed75591b1af2d675a240ea94c?permalink_comment_id=5119222#gistcomment-5119222

But, jemalloc do not call clock_gettime() that frequently (I've verified it), so it can use non-coarse version - CLOCK_MONOTONIC

I've also measured the latency of CLOCK_MONOTONIC and CLOCK_MONOTONIC_COARSE, and it is 20ns vs 4ns per call [3], so make this change affect performance you need really frequently calls of clock_gettime.

  [3]: https://gist.github.com/azat/622fa1f9a5d8e7d546ee9d294501961d?permalink_comment_id=5119245#gistcomment-5119245

Interesting, that this bug started to appears only after jemalloc heap profiler had been enabled by default [4], no clue why (I would believe more in a more frequent calls to clock_adjtime(ADJ_FREQUENCY), but I can't verify this)

  [4]: https://github.com/ClickHouse/ClickHouse/pull/65702

To be continued...

Fixes: https://github.com/ClickHouse/ClickHouse/issues/66193 (cc @antonio2368 )

azat committed:
Fix jemalloc assertion due to non-monotonic CLOCK_MONOTONIC_COARSE

Recently one tricky assertion of jemalloc had been discovered [1]:

    Failed assertion: "nstime_compare(&decay->epoch, new_time) <= 0"

  [1]: https://github.com/ClickHouse/ClickHouse/issues/66193

And as it turns out it is really possible for CLOCK_MONOTONIC_COARSE to
go backwards, in a nutshell it can be done with ADJ_FREQUENCY, you can
find example here [2]. And I can't trigger this issue for non-coarse
clocks.

  [2]: https://gist.github.com/azat/7ea7f50ed75591b1af2d675a240ea94c?permalink_comment_id=5119222#gistcomment-5119222

But, jemalloc do not call clock_gettime() that frequently (I've verified
it), so it can use non-coarse version - CLOCK_MONOTONIC

I've also measured the latency of CLOCK_MONOTONIC and
CLOCK_MONOTONIC_COARSE, and it is 20ns vs 4ns per call [3], so make this
change affect performance you need really frequently calls of
clock_gettime.

  [3]: https://gist.github.com/azat/622fa1f9a5d8e7d546ee9d294501961d?permalink_comment_id=5119245#gistcomment-5119245

Interesting, that this bug started to appears only after jemalloc heap
profiler had been enabled by default [4], no clue why (I would believe
more in a more frequent calls to clock_adjtime(ADJ_FREQUENCY), but I
can't verify this)

  [4]: https://github.com/ClickHouse/ClickHouse/pull/65702

To be continued...

Fixes: https://github.com/ClickHouse/ClickHouse/issues/66193
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

robot-ch-test-poll3 commented:
<!-- automatic status comment for PR #66439 from azat/ClickHouse:jemalloc-clock-assertion -->
*This is an automated comment for commit 32ef37996a1359bc79791bd653b61bb91cd4b000 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/66439/32ef37996a1359bc79791bd653b61bb91cd4b000/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/ad21a9307fc1d0715388923d8f1e490a7bd9bebe/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

antonio2368 commented:
Let's enable profiler back for debug so we don't have different configurations
https://github.com/ClickHouse/ClickHouse/blob/master/contrib/jemalloc-cmake/CMakeLists.txt#L37
Remove if else here

azat committed:
Enable jemalloc heap profiler by default for debug builds as well

Since the issue with jemalloc assertion is clear, let's revert that
workaround

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

antonio2368 commented:
If performance tests are happy then okay
If performance tests are not happy we will patch MONOTONI_OARSE to be `non-monotonic`.

alexey-milovidov commented:
@azat Let's submit a fix to the Linux man page. It says that
> All CLOCK_MONOTONIC variants guarantee that the time returned by consecutive calls will not go backwards, but successive calls may—depending on the architecture—return
              identical (not-increased) time values.

azat commented:
I haven't finished yet:

>To be continued...

By this I meant:
- jemalloc
- linux

alexey-milovidov commented:
Depends on https://github.com/ClickHouse/ClickHouse/pull/66460

azat commented:
>Stateless tests (ubsan) — fail: 1, passed: 6692, skipped: 57

- #66107 

>Stateless tests (debug, s3 storage) [1/2] — fail: 1, passed: 3296, skipped: 69

- `01069_window_view_proc_tumble_watch`

>Stateful tests (tsan) — Timeout 3600 exceeded

- https://github.com/ClickHouse/ClickHouse/pull/66036#discussion_r1677150007
- https://github.com/ClickHouse/ClickHouse/pull/65045#discussion_r1677151338

alexey-milovidov commented:
@azat, your pull request is mistakenly classified as a bugfix, but it does not fix any wrong behavior in production ClickHouse builds. It does not even have tests to prove that. Our CI pointed to that. Please change the classification.

antonio2368 commented:
I'm going to merge as test failures seem unrelated and performance tests are okay
